libmisc/csrand.c

```diff
@@ -59,3 +65,23 @@ csrand(void)
 	fprintf(log_get_logfd(), _("Unable to obtain random bytes.\n"));
 	exit(1);
 }
+
+
+/*
+ * Return a uniformly-distributed CS random value in the interval [0, n-1].
+ */
+unsigned long
+csrand_uniform(unsigned long n)
+{
+	unsigned long  r, max, mask;
+
+	max = n - 1;
+	mask = bit_ceil_wrapul(n) - 1;
+
+	do {
+		r = csrand();
+		r &= mask;  // optimization
+	} while (r > max);
```

Comment:
From lack of knowledge, how long would it take on average to execute this loop? In other words, is it possible that it would take a long time to execute and that the user experience would suffer?

Reply:
Hmm, I'll add some documentation in the code clarifying that.

It's a [geometric distribution](https://en.wikipedia.org/wiki/Geometric_distribution), where `p` is somewhere between `0.5` and `1`, depending on the value of `max` (if `powerof2(max + 1) == 1`, `p == 1`; if `powerof2(max) == 1`, then `p` is close to `0.5`).

A geometric distribution, theoretically, can loop forever; however, that's very unlikely (the probability of that is an infinitesimal).

Being realistic, and assuming a worst case (the worst case is for `csrand_uniform(LONG_MAX + 2ul)`, so `max == LONG_MAX + 1ul`, so `p` ~= `0.5`), let's get some numbers:

Pr(X = k) = (1 - p)^(k - 1) * p ~= 0.5^(k - 1) * 0.5 = 0.5^k

Now, let's assume that every human in the world (around 2^33) would need a million random numbers per day (around 2^20) for 10000 years (around 2^22 days), we see that k (the number of iterations in the loop) might reach 75 some day in the next 10k years.

However, that's just for seeing what is likely to happen some day on earth.  On average (even for a worst case input), it's likely to loop once, or just a few times.  It can never be worse than tossing a coin and waiting for the first heads.

Reply:
I'm still trying to understand why the multiplication that Jason suggested is an optimization.  If you want me to do that on this PR, we can wait.